### PR TITLE
[JENKINS-58767] Persist yaml fragment in a single string

### PR DESCRIPTION
**Pull request by Vlatombe**
[JENKINS-58767](https://issues.jenkins-ci.org/browse/JENKINS-58767)

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java [hunk 1]
```diff
@@ -716,18 +722,16 @@ protected Object readResolve() {
...
-        // JENKINS-57116 remove empty items from yamls
-        if (!yamls.isEmpty() && StringUtils.isBlank(yamls.get(0))) {
+        // Sanitize empty values
+        yaml = Util.fixEmpty(yaml);
+        if (yamls != null) {
+            // JENKINS-57116 Sanitize empty values
             setYamls(yamls);
+            // Migration from storage in yamls field
+            if (!yamls.isEmpty()) {
+                yaml = yamls.get(0);
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: Should this not be verifying that `yamls.size() == 1`?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java [hunk 2]
```diff
@@ -637,27 +637,32 @@ public void setContainers(@Nonnull List<ContainerTemplate> items) {
      */
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: > The first yaml fragment for this pod template

Is that still right?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java [hunk 3]
```diff
@@ -123,12 +124,14 @@
...
-     * @deprecated Stored as a list of yaml fragments
+     * Persisted yaml fragment
      */
-    @Deprecated
-    private transient String yaml;
+    private String yaml;
 
-    private List<String> yamls = new ArrayList<>();
+    /**
+     * List of yaml fragments used for transient pod templates. Never persisted
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: Could this field just be deleted, and any call to `setYamls` with a list of size > 1 would just set `yaml` to something separated by `"\n----\n"` or whatever? Would be easier to follow the logic and be sure that you are not missing some complication.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java (hunk 1 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: Yes, could even be asserted I guess (unless someone tampered manually with the xml)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java (hunk 3 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: Problem I see is that would end up being an invalid yaml fragment to parse into a pod template

Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: As you see fit. I do not really understand what the list of YAMLs is for to begin with.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesTest.java [hunk 4]
```diff
@@ -96,6 +101,19 @@ public void upgradeFrom_1_15_9() {
         assertNull(template._getYamls());
     }
 
+    @Test
+    @LocalData
+    public void upgradeFrom_1_15_9_invalid() {
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesTest.java]: Probably overkill, but OK.

**Pull request merged**